### ISSUE TITLE
`causal` argument when q_sequence is not `cu_seqlens_q` isn't aligned with `cu_seqlens_k`

### ISSUE THREAD

**Issue by thomasw21** [HazyResearch/flash-attention#282]
First of, thanks for making this amazing repo available to everyone! We're using this repository on a daily basis, and it allows us to achieve impressive throughput.

We noticed that the use of `causal` argument is a bit weird whenever the sequences in `k` are longer than `q` (typically this happens in decoding). I'm not entirely sure but I'm guessing that when `causal=True`, the first query attends to the first key until one of them runs out, the second query to the second key and the first key. Instead I would have expected to have the last query attends to all keys, the previous query attends to all keys except the last one.

Note this is pure speculation, I'm not yet familiar with the implementation so I might be completely off.

Example: in https://github.com/huggingface/text-generation-inference: 
 - we first process the prompt as a whole using `causal=True`: https://github.com/huggingface/text-generation-inference/blob/c9c65ab323f48731e1fc2f7087547a7bd8b753f2/server/text_generation_server/models/custom_modeling/flash_santacoder_modeling.py#L195
 - Once we start decoding using single token queries, we deactivate that flag: https://github.com/huggingface/text-generation-inference/blob/c9c65ab323f48731e1fc2f7087547a7bd8b753f2/server/text_generation_server/models/custom_modeling/flash_santacoder_modeling.py#L224
 
cc @OlivierDehaene

**Comment by tridao** [comment]
Yes your understanding is right.
For [inference](https://github.com/HazyResearch/flash-attention/blob/9818f85fee29ac6b60c9214bce841f8109a18b1b/flash_attn/modules/mha.py#L503), I set `causal=True` for prompt processing and `causal=False` during iterative decoding where the query length is 1.

**Comment by thomasw21** [comment]
Would it make sense to change the behaviour to match trailing tokens? Last query can attends all keys, the one before can attends all keys except last etc ...

This would remove the switch in causal when decoding. Also it would allow to batch inputs that are at different step of decoding.

**Comment by tridao** [comment]
I haven't seen this use case very much, possibly because I'm not super experienced in LM inference. 
I actually don't use FlashAttention for iterative decoding (only for training/fine-tuning and prompt-processing), so for best inference performance there needs to be some case work anyway.

**Comment by thomasw21** [comment]
We've been using flash attention for inference and it works quite well. We just have this switch that's slightly annoying. Happy to have a go at it if you agree on the need.

**Comment by tridao** [comment]
I'd prefer not to change this behavior of `causal=True` for backward compatibility and ease of implementation (in CUDA) if the purpose is to avoid some case work in Python.

**Comment by thomasw21** [comment]
Fair enough!

**Comment by masahi** [comment]
Hi @thomasw21, I'm seeing an odd output from flash (and also `scaled_dot_product_attention`) when it is applied to decoder attention with KV cache.  

When `is_causal = True` and there is only a single query, the output of both flash and PT 2.0 SDP don't seem correct. Does this agree with your observation?

**Comment by tridao** [comment]
This is now implemented. The behavior is changed so that the causal mask is aligned to the right.

**Comment by fangzhangmnm** [comment]
> This is now implemented. The behavior is changed so that the causal mask is aligned to the right.

both the document and the code gives the left-aligning result. I've already updated pytorch to 2.5.0. how to toggle on the new right-aligning behavior?
